man/sendfile.1
.\" $Header$
.\"
.\"  Man page for HUJI-NJE on UNIX system.
.\"
.\"  Written by  Matti Aarnio <mea@finfiles.bitnet> <mea@nic.funet.fi>
.\"
.\"  Date: 27-Dec-1990, 20-Oct-1993
.\"
.ds ]W Funet-Nje 1 UNIX 1.1
.TH SENDFILE 1
.SH NAME
sendfile, submit, punch, print \- bitnet file send tools
.SH SYNOPSIS
.IP sendfile 0.8i 0
.I ToUser@ToNode
.RB [ " \-fix | \-raw " ]
.RB [ " \-bin " ]
.RB [ " \-lrecl nn " ]
.RB [ " \-fn \fIfname ftype " ]
.RB [ " \-class N " ]
.RB [ " \-form \fIFromName " ]
.RB [ " \-dist \fIDistribution " ]
.RB [ " \-tag \fIRSCS-TAG " ]
.RI [ filepath ]
.IP punch 0.8i 0
.I ToUser@ToNode
.RB [ " \-rawpun " ]
.RB [ " \-fix | \-raw " ]
.RB [ " \-fn \fIfname ftype " ]
.RB [ " \-class N " ]
.RB [ " \-form \fIFromName " ]
.RB [ " \-dist \fIDistribution " ]
.RB [ " \-tag \fIRSCS-TAG " ]
.RI [ filepath ]
.IP print 0.8i 0
.I ToUser@ToNode
.RB [ " -asa " ]
.RB [ " \-fix | \-raw " ]
.RB [ " \-bin " ]
.RB [ " \-fn \fIfname ftype " ]
.RB [ " \-class N " ]
.RB [ " \-form \fIFromName " ]
.RB [ " \-dist \fIDistribution " ]
.RB [ " \-tag \fIRSCS-TAG " ]
.RI [ filepath ]
.IP submit 0.8i 0
.I ToUser@ToNode
.RB [ " \-fix | \-raw " ]
.RB [ " \-form \fIFromName " ]
.RI [ filepath ]
.SH DESCRIPTION
.IX "bitnet"
.IX "funetnje"
.IX "sendfile"
.IX "submit"
.IX "netdata"
.LP
.I Sendfile
command sends out
.SM NETDATA
encoded file
with given
.I "fname + ftype"
and
.I "FormName,"
if they are specified.
.LP
.I Punch
command sends out 80-chars per record punch-card images
with given
.I "fname + ftype"
and
.I "FormName,"
if they are specified.
.LP
.I Submit
command sends out a punch with special flags directing it to be
a NJE SYSIN job.
.LP
When a file is defined,
.I reallylongfilename.really-long-extension.more.dots
maps to:
.I "reallylongfile + really-long-"
(12 + 12 chars.)
Default filename (when nothing is defined) is:
.IR "STANDARD INPUT" .
.LP
Default for
.I form
is
.IR STANDARD .
.LP
When no input file is defined,
.I STDIN
is assumed
.LP
Options for
.BR punch ,
.BR print ,
.BR sendfile ,
and 
.B submit
are:
.IP "\fB\-asa\fR" 1.8i 0
For
.IR print ", and " sendfile .
A print format which expects to get
.SM FORTRAN
carriage control style input.
.IP "\fB\-bin\fR" 1.8i 0
For
.BR sendfile (1)
only.
Defines binary transmission, and defaults class `N', if nothing
else is defined.
.IP "\fB\-class C\fR" 1.8i 0
Defines single uppercase alpha letter for the NJE spool
class of the file.   For
.B sendfile
the
\-bin
option defaults class to be `N', otherwise default is `A'.
.IP "\fB\-fix\fR" 1.8i 0
Makes padded records of native width; 80 for punch, and submit,
132+1 for print.
On sendfile it asks output to be of fixed size.  On such cases
input lines get padded by the receiver to the fixed format.
.IP "\fB\-fn fname ftype\fR" 1.8i 0
For
.BR sendfile "(1), " punch "(1), " print "(1) only."
Defines NJE level \fI`file name'\fR and \fI`file type'\fR strings.
Default values are: `STANDARD', and `INPUT'.
.IP "\fB\-form FormName\fR" 1.8i 0
Defines NJE level \fI`file form name'\fR to be used.
Default value is `STANDARD'.
.IP "\fB-lrecl ##\fR" 1.8i 0
For
.BR sendfile (1)
defines logical record length.
Defaults are: binary: 1022, others: 132.
If your ASCII input is wider (or narrower), do consider using this
option to optimize receiver performance.
.IP "\fB\-rawpun\fR" 1.8i 0
For
.BR punch (1)
only.
Outputs fixed 80 chars per record lines, and expects same from input.
(No cr/lfs wanted, just 80 bytes for each record.)
.IP "\fB\-dist Distribution\fR" 1.8i 0
Defines the RSCS Distribution code (max. 8 characters) which may be
used to identify printed outputs. Defaults to user name of the sender
translated to upper case. If not set here, the emulator inserts the
default SYSTEM.
.IP "\fB\-tag RSCS-TAG\fR" 1.8i 0
Defines NJE RSCS TAG (136 characters) which is used to carry various
information especially at RJE applications.
.LP
Default behaviour is to read in lines, and convert them from
.SM ASCII
to
.SM EBCDIC
and pad or truncate to 80 char (132 for print) records
which get spooled for transmission.
.LP
.SH ENVIRONMENT
With environment variable
.B EBCDICTBL
user can select alternate EBCDIC <\-> ASCII translation
tables.  For more info, see
.BR ebcdictbl (5).
.LP
.SH FILES
.I /etc/funetnje.cf
.SH SEE ALSO
.BR bitspool (5),
.BR bmail (8),
.BR ebcdictbl (5),
.BR funetnje (8),
.BR mailify (8),
.BR qrdr (8),
.BR receive (1),
.BR send (1),
.BR transfer (1)
.SH BUGS
.LP
Sendfile should figure out the lrecl in text files by smart pre-scanning
of the file -- doable if sending real files ?
.LP
.SH DIAGNOSTICS
These programs show their usage summary, when there is something
wrong with input parameters.  Subtle usage bugs might not be detected!
.SH NOTES
This program quartet is actually same program with 4 different
invocation names, and therefore 4 (slightly) different functions.
.SH AUTHORS
.LP
Yehavi Bourvine, Hebrew University of Yerusalem, Israel
.I <yehavi@hujivms.bitnet>.
.LP
Matti Aarnio, Finnish University and Research Network \- FUNET, Finland
.I <mea@nic.funet.fi> <mea@finfiles.bitnet>.
.LP
Gerald Hanusch, Johannes Kepler Universitaet/Universitaet Linz, Austria
<k000165@alijku11.bitnet>.
.LP
Original (VMS oriented package) software by Yehavi Bourvine,
UNIX features, plenty of modifications, file handling exits,
man-pages by Matti Aarnio.
Batch (SYSIN) facilities, and bug fixes prompted by Gerald Hanush,
and in large part implemented by Matti Aarnio.


ChangeLog
Fri Mar 22 13:48:30 1996  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    Even more relaxed approach on parametrizing
	    the CMDMAILBOX -- now allows SOCK/FIFO/UDP
	    with fairly flexible (in regards of spaces)
	    syntax..

Mon Nov 13 17:48:52 1995  Matti Aarnio  <mea@nic.funet.fi>

	* libsendcmd.c, unix.c:
	    A bit more relaxed approach on parametrizing
	    CMDMAILBOX config parameter (FIFO/SOCKET case)

	* file_queue.c:
	    Do RSCS-accounting on locally sent files, that
	    wind up to local recipients

	* unix_brcst.c:
	    Re-enter the SysVR4 UTMP weirdos..  Now fixed ?

Fri Aug 25 09:02:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* mailify.c:
	    Handling error-messages (where "MAIL FROM:" address
	    is "<>", that is, empty) was done improperly.  Now
	    it does flag input as from "channel error" in that
	    case.

Mon Jul 24 14:29:59 1995  Matti Aarnio  <mea@nic.funet.fi>

	* bmail.c, sendfile.c:
	    Sometimes   mktemp()  fails when there are a LOT of
	    files being sent at the same time.  Changed now to
	    more classical  ".MaiL.%05d"-format with  getpid()
	    result as integer data..

Wed Jun 21 22:42:39 1995  Matti Aarnio  <mea@nic.funet.fi>

	* unix_brcst.c, nmr_unix.c:
	    Ok, more than __linux__ is needed to know which
	    machines have <utmp.h> containing extended flags.
	    Now we test  "DEAD_PROCESS" pre-processor define,
	    which existence (propably) will correlate with
	    the existence of utmp's field  ut_type.
	    (SysVr4 feature, I think.)

Mon Jun 19 15:17:55 1995  Matti Aarnio  <mea@nic.funet.fi>

	* send.c, sendfile.c, bmail.c:
	    Some additional tests to trap various forms of bad
	    addresses -- accepts only  user@node format with
	    no '.' in the address, nor either component longer
	    than 8 chars.

	* nmr.c:
	    Made sure not to overwrite allocated string spaces on
	    Msg-queue entry.  (Did cause a crash with it..)

Tue Jun 13 09:13:18 1995  Matti Aarnio  <mea@nic.funet.fi>

	* qrdr.c:
	    Collect directory browsing results to an array at first,
	    then qsort() it with the time-stamp, and finally display
	    them in order. (Latest last)
	    (This is something what Gerald Hanush asked for.)

Mon Jun 12 22:44:00 1995  Matti Aarnio  <mea@nic.funet.fi>

	* file_queue.c:
	    Apparently OSF/1 pipe() return value is the first
	    fd, not "0", when it succeeds.  Nice thing - and not..
	    Altered the test to see if it returns a value under
	    zero, or something positive.  When the return is
	    positive (zero, or larger), fork() and run in
	    child-mode.

	    If a child_pipe exists already ( >= 0 ), close
	    it and mark the new one for processing.

	* libreadcfg.c, read_config, libsendcmd.c, unix.c:
	    Rewrote things so that now the used CMDMAILBOX
	    transport method is chosen at runtime:

	    CMDMAILBOX  F  /path/to/.cmdfifo
	    CMDMAILBOX  f  /path/to/.cmdfifo
	    CMDMAILBOX  S  /path/to/.cmdsocket
	    CMDMAILBOX  U  127.0.0.2 175	# UDP port

	    The "f" is FIFO which is closed, and reopened
	    every time one command is pulled in.

Sun Jun 11 03:12:06 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesparser.c:
	    Noticed that the data parsing stopped at first ":",
	    which caused premature breaks; observe following entry
	    from BITEARN.NODES -file:
	    	:remark1.BASE(9501) 00/00/00 00:00:00

	* libreadcfg.c, read_config.c, libsendcmd.c, unix.c:
	   Amended "CMDMAILBOX"-data to contain port-number in
	   case UDP-format is used. (That is, have two tokens.)
	   Default value for the port-number is 175.

	* unix_tcp.c read_passive_tcp_connection():
	   Noticed that a case when a socket exists, and a new
	   connection is attempted is handled improperly. The
	   old socket-fd was FORGOTTEN!  Now it NAKs the connection
	   attempt with a complaint that there already is a connection
	   in there...

	   Things will progress from such a situation usually by
	   detecting a closure on the existing stream, but better
	   not to rush ahead of things.

	* bmail.c:
	   Accept "-u" only when calling the program from uids 0 or 1
	   (root and daemon in most common cases -- for BSD-sendmail
	    which wants  bmail executable with protection 2111 (user X,
	    group NJEGRP) )

Sat Jun 10 05:18:27 1995  Matti Aarnio  <mea@nic.funet.fi>

	* Makefile, consts.h, prototypes.h, recv_file.c, io.c,
	  send_file.c, unix.c, protocol.c, unix_tcp.c:

	    - Found one uninitialized pointer, which apparently
	      caused mystic data-structure corruptions (on timer
	      subsystem)

	    - Made mandatory (well, with disabling options) to
	      use NBSTREAM, USE_XMIT_QUEUE, and USE_SOCKOPT.
	      (in consts.h, see also  Makefile's  comments!)

	    - Use  F_WAIT_V_A_BIT  to flag if there is any reason
	      for NOT to send out data at the moment, and instead
	      quit from within  handle_ack()..
	      The F_WAIT_V_A_BIT is set when new data is sent to
	      outbound buffer (primary one, or XMIT QUEUE).
	      The bit is cleared when pending writer finishes
	      with success status.  When the pending-writer is
	      called from  poll_sockets(), immediately after it
	      the  dequeue_xmit_queue() is called, and if there
	      is something, it can cause F_WAIT_V_A_BIT to be
	      set again..)

	    - If the XMIT_QUEUE ever runs out of space, it abort()s!

	    - Tested the beast by sending 80 or so smallish files both
	      ways over the link.  Such setup caused previously link
	      corruption..  Actually the critical phase is when there
	      are flowing NJE headers/trailer, bare data sending works
	      just fine, but when a file ends, EOF block needs an ack
	      to be sent, and unless there the care is used, all
	      queues will overflow..

Thu Jun  8 15:48:14 1995  Matti Aarnio  <mea@nic.funet.fi>

	* namesfilter.c:
	    New option: "-netsoft" to list node software fields.

	* funetnje.texinfo:
	    Some additional chapters

	* unix_tcp.c:

sendfile.c
/*   SENDFILE.C	V1.5-mea
 |
 | Copyright (c) 1990,1991,1993 by
 | Finnish University and Research Network, FUNET
 |
 | Calling sequence:  SENDFILE - see usage()...
 |
 | V1.0 [mea]  Modified BMAIL to SENDFILE... More is needed...
 |		(like NETDATA et.al. processings)
 | V1.1 [mea] - 9-May-91
 |		Added somehow working -print, but no -fortran options done...
 | V1.2 [mea] - 5-Jul-91
 |		Plenty of bug fixes. Getting BITNET mail to work.
 | V1.3 [mea] - 18-Sep-93
 |		RSCS TAG, and some cleanups.
 | V1.4 [mea] - 9-Oct-93
 |		Integrating netdata generation code into this module.
 | V1.5 [mea] - 10-Oct-93
 |		Start-name affects operation mode:
 |			punch, print, submit, sendfile
 */


#include "prototypes.h"
#include "clientutils.h"
#include "ndlib.h"
#include <sysexits.h>

#define	LSIZE	256			/* Maximum line length */

extern char DefaultForm[9];
char LOCAL_NAME   [10];
char BITNET_QUEUE [80];
char LOG_FILE     [80] = "-";

int LogLevel = 1;
FILE *LogFd = NULL;

enum opmodes { UnKnown, Punch, Sendfile, Print, Submit };
enum opmodes OpMode = UnKnown;

char *cmdname = "<sendfile>";

/* ================ usage() ================ */

void
usage(explain)
char *explain;
{
  fprintf(stderr, "punch/print/sendfile/submit:\n");
  fprintf(stderr, "Usage: %s ToUser@ToNode [-u origuser] [-class SpoolClass]\n",cmdname);
  fprintf(stderr, "            [-fix|-var]");
  if (OpMode == Sendfile)
    fprintf(stderr, " [-bin]");
  if (OpMode != Submit)
    fprintf(stderr, " [-fn fname ftype] [-form FormName]\n            [-dist Distribution] [-tag RSCS-TAG]");
  if (OpMode == Print || OpMode == Sendfile)
    fprintf(stderr, " [-asa]");
  if (OpMode == Punch)
    fprintf(stderr, " [-rawpun]");
/* Sendfile: -lrecl !! */
  fprintf(stderr, " [filepath]\n");
  if (explain) {
    fputs(explain,stderr);
    fputc('\n',stderr);
  }
  exit(EX_USAGE);
}


/* ================ main() ================ */

int
main(argc, argv)
int	argc;
char	*argv[];
{
	unsigned char	From[LSIZE], To[LSIZE], 
			fname[20], ftype[20], Forms[9], Dist[9], FrUser[9];
	char	FileName[LSIZE], tFileName[LSIZE];
	char	FileClass = 0;
	int	binary = 0;
	long	reccount = 0;
	int	len, i;
	unsigned char	ebcdicline[132+2+3];
	unsigned char	outbuf[150];
	unsigned char	line[1000], *p;
	FILE	*fd, *infile;
	int	rawpun = 0;
	int	asaform = 0;
	int	fnametypeset = 0;
	struct stat stats;
	char	*Tag = NULL;
	char	*AskAck = NULL;
	int	FileSize = 0;
	int	rc;
	long	recpos = 0;
	int	fixedwidth = 0;
	int	lrecl = 0;
	int	recfm = ND_VARY; /* Alternate: ND_FIXED ?? */

#if 0 /* Some temporary debug code.. */
{ char **aa = argv;
  fprintf(stderr,"sendfile: argc=%d, argv=[\"%s\"",argc,*aa++);
  for (; *aa; ++aa) fprintf(stderr,",\"%s\"",*aa);
  fprintf(stderr,"]\n");
}
#endif

	cmdname = strrchr(*argv,'/');
	*To = 0;

	if (cmdname == NULL) cmdname = *argv;
	else ++cmdname;
	if (strcmp(cmdname,"sendfile") == 0) {
	  OpMode = Sendfile;
	} else if (strcmp(cmdname,"sf") == 0) {
	  OpMode = Sendfile;
	} else if (strcmp(cmdname,"print") == 0) {
	  OpMode = Print;
	} else if (strcmp(cmdname,"bitprt") == 0) {
	  OpMode = Print;
	} else if (strcmp(cmdname,"punch") == 0) {
	  OpMode = Punch;
	} else if (strcmp(cmdname,"submit") == 0) {
	  OpMode = Submit;
	} else {
	  OpMode = UnKnown;
	  fprintf(stderr,"`sendfile' invoked with name (%s) that is not proper OpMode for it!\n",cmdname);
	  exit(EX_USAGE);
	}

	*From = 0;
	*Dist = 0;
	memset( ebcdicline, 0, sizeof ebcdicline );

	if (argc < 2) usage(NULL);

	strcpy( fname,"STANDARD" );
	strcpy( ftype,"INPUT"    );

	/* Read our name and the BITnet queue directory */
	read_configuration();
	read_etable();

	strcpy( Forms, DefaultForm);

	++argv;
	--argc;

	/* ================ Options ================ */

	while (*argv) {

	  /*fprintf(stderr,"Checking option: `%s'\n",*argv);*/

	  if (strcmp (*argv,"-rawpun") ==0 ) {

	    if (OpMode != Punch)
	      usage("  Bad usage of `-rawpun'");
	    rawpun = 1;

	  } else if (strcmp(*argv, "-asa") == 0) {

	    if (OpMode != Print && OpMode != Sendfile)
	      usage("  Bad usage of `-asa'");
	    asaform = 1;

	  } else if (strcmp(*argv, "-fix") == 0) {

	    fixedwidth = 1;
	    recfm = ND_FIXED;	/* For the NETDATA format */

	  } else if (strcmp(*argv, "-var") == 0) {

	    fixedwidth = 0;

	  } else if (strcmp(*argv, "-class") == 0) {

	    ++argv;
	    if (!*argv)
	      usage("  -class option requires ONE parameter!");
	    --argc;
	    upperstr(*argv);
	    FileClass = **argv;

	  } else if (strcmp(*argv, "-fn") == 0) {

	    ++argv;
	    if (OpMode == Submit)
	      usage("  -fn is improper option for SUBMIT!");
	    if (!*argv || !*(argv+1))
	      usage("  -fn option requires TWO parameters!");
	    strncpy(fname, *argv,sizeof fname -1);
	    strncpy(ftype, *++argv,sizeof ftype -1);
	    fnametypeset = 1;
	    argc -= 2;

	  } else if (strcmp(*argv, "-form") == 0) {

	    ++argv;
	    if (OpMode == Submit)
	      usage("  -form is improper option for SUBMIT!");
	    if (!*argv)
	      usage("  -form  option requires one parameter!");
	    strncpy(Forms,*argv,sizeof Forms -1);
	    argc -= 1;

	  } else if (strcmp(*argv, "-dist") == 0) {

	    ++argv;
	    if (OpMode == Submit)
	      usage("  -dist is improper option for SUBMIT!");
	    if (!*argv)
	      usage("  -dist  option requires one parameter!");
	    strncpy(Dist,*argv,sizeof Dist -1);
	    argc -= 1;

	  } else if (strcmp(*argv,"-tag")==0) {

	    if (OpMode == Submit)
	      usage("  -tag is improper option for SUBMIT!");
	    if (!*++argv)
	      usage("  -tag  option requires one parameter!");
	    Tag = *argv;
	    argc -= 1;

	  } else if (strcmp(*argv,"-u")==0) {

	    char *s;

	    if (!*++argv)
	      usage("  -u  option requires one parameter!");
	    --argc;
	    strncpy(From,*argv,sizeof(From)-1);
	    From[sizeof(From)-1] = 0;
	    if ((s = strchr(From,'@')) == NULL) {
	      if (strlen(From) > 8)
	    fromaddrerror:
		usage("  -u option defined username components must be at most 8 chars long!");
	      strcat(From,"@");
	      strcat(From,LOCAL_NAME);
	    } else {
	      *s = 0;
	      if (strlen(From) > 8)
		goto fromaddrerror;
	      *s = '@';
	      if (strlen(s+1) > 8)
		goto fromaddrerror;
	    }
	    if (strchr(From,'.') != NULL)
	      usage(" -u -defined source address may not contain '.' chars in it");
	  } else if (strcmp(*argv,"-lrecl")==0) {

	    if (OpMode != Sendfile)
	      usage("  -lrecl  is usable only with  sendfile!");
	    if (!*++argv)
	      usage("  -lrecl  option requires one parameter!");
	    --argc;
	    if ((lrecl = atoi(*argv)) < 10 || lrecl > 65535)
	      usage("  Bad value of  -lrecl  parameter!  Must be within 10 - 65535.");

	  } else if (strcmp(*argv,"-bin") == 0) {

	    if (OpMode != Sendfile)
	      usage("  -bin  is usable only with  sendfile!");

	    binary = 1;
	    if (FileClass == 0)
	      FileClass = 'N';
	    if (lrecl == 0)
	      lrecl = 8192; /* BINARY is wider.. */

	  } else if (**argv != '-' && *To == 0) {
	    char *s;
	    strncpy(To, *argv, sizeof(To)-1);
	    if ((s = strchr(To,'@'))==NULL) {
	      if (strlen(To) > 8) {
	    targetaddrerror:
		usage(" Target address has components of more than 8 chars long");
	      }
	      /* To[8] = 0; */
	      sprintf(To+strlen(To),"@%.8s",LOCAL_NAME);
	    } else {
	      *s = 0;
	      if (strlen(To) > 8)
		goto targetaddrerror;
	      *s = '@';
	      if (strlen(s+1) > 8)
		goto targetaddrerror;
	    }
	    if (strchr(To,'.') != NULL)
	      usage(" Target address may not contain '.' in its components");
	  } else
	    break;

	  ++argv; --argc;
	} /* While options.. */


	if (*To == 0) {
	  usage("   Missing recipient address!");
	}

	if (FileClass == 0)
	  FileClass = 'A';

	infile = stdin;
#if 0 /* Some ARGV processing debug code */
fprintf(stderr," figuring source file,  *argv='%s'\n",*argv);
#endif
	if (*argv) { /* extra file path... ? */
	  if ((infile = fopen(*argv,"r")) == NULL) {
	    fprintf(stderr,"Can't open `%s' -file for reading!\n",*argv);
	    exit(3);
	  }
	  /* fclose(stdin); */
	  /* Don't close it! There are surprises around
	     to wait it..  Freshly open file may disappear.. */

	  if (!fnametypeset) {
	    /* No fname+ftype set yet (aside of default), set here */
	    /* Find base name - after last `/' that is. */
	    char *fna = strrchr(*argv,'/');
	    char *s, *s2;
	    if (!fna)
	      fna = *argv;
	    else
	      ++fna;

	    /* Locate the first `.' in it */
	    s = strchr(fna,'.');
	    s2 = s;
	    while (s2) {
	      /* And turn other `.'s to `-'es.. */
	      s2 = strchr(s2+1,'.');
	      if (s2) *s2 = '-';
	    }
	    /* And chop them to pieces.. */
	    if (s) {
	      *s = 0;
	      ++s;
	    }
	    else
	      s = "";
	    /* fna points to fname part, s to ftype part */
	    strncpy(fname, fna, sizeof fname -1);
	    fname[12] = 0;
	    strncpy(ftype, s, sizeof ftype -1);
	    ftype[12] = 0;
	  }
	  ++argv;
	}
	if (*argv) usage(" extra arguments/unknown options?");
	

	if (getuid() >= LuserUidLevel) {
	  if (*From) {
	    fprintf(stderr," -u option can't be used unless sufficiently priviledged!\nYour UID=%d\n",(int)getuid());
	    exit(9);
	  }
	  *From = 0;
	}
	/* Not yet set, set it ! */
	if (*From == 0) {
	  if (mcuserid(From) == NULL) {
	    fprintf(stderr,"Can't determine who you are. Aborting!\n");
	    exit(2);
	  }
	  strncpy(FrUser,From,8);
	  FrUser[8] = 0;
	  strcat(From,"@");
	  strcat(From,LOCAL_NAME);
	}

/* Upcase From and To to prevent problems with RSCS */

	upperstr(fname);
	upperstr(ftype);
	upperstr(From);
	upperstr(To);
	upperstr(FrUser);

#define CHOP_BITNET(VarName) \
	i = strlen((VarName)); \
	if (i > 8 && strcasecmp((VarName)+i-7,".BITNET") == 0) \
	  (VarName)[i-7] = 0
	CHOP_BITNET(From);
	CHOP_BITNET(To);


	/* Ok, SGID program to send files... */
	setgid(getegid());


	sprintf(tFileName, "%s/.BIT_%05d", BITNET_QUEUE, getpid());
	if (*tFileName == '\0') {
		fprintf(stderr, "Can't create unique filename\n");
		exit(1);
	}

	umask(077); /* THIS IS NOT A PUBLIC FILE! */
	if ((fd = fopen(tFileName, "w")) == NULL) {
		perror(tFileName); exit(1);
	}
	fstat(fileno(fd),&stats);

	
	/* The RSCS envelope for our NJE emulator */

	/* Form up the initial block */
	fprintf(fd, "%-506s\nEND:\n","*");
	fflush(fd);
	fseek(fd,0,0);

	fprintf(fd, "FRM: %-17s\n", From);
	fprintf(fd, "FMT: BINARY\n");
	if (OpMode == Submit)
	  fprintf(fd, "TYP: JOB\n");
	else if (OpMode == Print)
	  fprintf(fd, "TYP: PRINT\n");
	else	/* OpModes  Sendfile, and Punch look similar.. */
	  fprintf(fd, "TYP: PUNCH\n");
	fprintf(fd, "FNM: %-12s\nEXT: %-12s\n",fname,ftype);
	fprintf(fd, "TOA: %-17.17s\nFOR: %-8s\nCLS: %c\n", To,Forms,FileClass);
	fprintf(fd, "FID: 0000\nOID: 0000\n");
	if (*Dist != 0)
	  fprintf(fd, "DIS: %-8s\n", Dist);
	  else
	  fprintf(fd, "DIS: %-8s\n", FrUser);
	if (Tag)
	  fprintf(fd, "TAG: %-136.136s\n",Tag);
	recpos = ftell(fd);
	fprintf(fd, "REC: %8d",1); /* NO \n ! */
	fflush(fd);

	fseek(fd,0,2); /* To the end.. */

	/* The message itself - copy it as-is */

	if (OpMode == Sendfile) {

	  /* ================ Sendfile ================ */

	  struct puncher	PUNCH;

	  PUNCH.buf[0]   = CC_NO_SRCB;
	  PUNCH.buf[1]   = 80;
	  PUNCH.fd       = fd;
	  PUNCH.len      = 0;
	  PUNCH.punchcnt = 0;

	  if (lrecl == 0)
		lrecl = 132;	/* set default lrecl */	

	  if (!do_netdata(infile,&PUNCH,From,To,
			  fname,ftype,lrecl,recfm,
			  binary,AskAck)) {
	    unlink(tFileName);
	    fprintf(stderr," write of bitnet spool file failed!\n");
	    exit(EX_TEMPFAIL);
	  }

	  reccount += PUNCH.punchcnt;

	} else if (OpMode == Punch && rawpun) {

	  while ((len = fread(line,1,80,infile)) != 0) {
	    if (len <80)	/* Pad with zero */
	      memset(line+len,0,80-len);
	    outbuf[0] = 0x80;
	    outbuf[1] =   80;
	    memcpy( outbuf+2,line,80 );
	    if (!Uwrite( fd,outbuf,82 )) {
	      unlink(tFileName);
	      fprintf(stderr,"PUNCH/rawpun: write of bitnet spool file failed!\n");
	      exit(EX_TEMPFAIL);
	    }
	    ++reccount;
	  }

	} else {

	  /* ================ Punch, Print, Submit ================ */

	  int maxlen = 80;
	  int rectype = CC_NO_SRCB;
	  int cc_space = 0;

	  if (OpMode == Print) {
	    maxlen = 132;
	    rectype = CC_MAC_SRCB;
	    cc_space = 1;	/* +1 for carriage control */
	    
	    if (asaform) {	/* PRINT ASA format */	
	      maxlen = 133;
	      rectype = CC_ASA_SRCB;
	      cc_space = 0;	/* +1 for carriage control */
	    }
	  }

	  while (fgets(line, sizeof line, infile) != NULL) {

	    len = strlen(line);

	    p = line + len;
	    if(len>0)		/* Zap the \n! */
	      if (*(p-1)=='\n')
		--len;
	    if (len==0) { *line=' '; len=1; } /* At least 1 byte! */
	    if (len > maxlen) len = maxlen;
	    if (fixedwidth)
	      for (;len < maxlen; ++len)
		line[len] = ' ';

	    ebcdicline[2] = rectype;
	    if (asaform)
	      ebcdicline[3] = maxlen-1;
	    else
	      ebcdicline[3] = maxlen;
	    p = ebcdicline + 4;
	    if (OpMode == Print && !asaform) {
	      /* We know NOTHING about carriage control! */
	      *p++ = 0x09;	/* Until we know, lets put 0x09 */
	    }
	    ASCII_TO_EBCDIC( line,p,len );
/*	    if (OpMode == Print && asaform) {
	      --len;
	    } */
	    if (!Uwrite( fd, ebcdicline+2, 2+cc_space+len )) {
	      unlink(tFileName);
	      fprintf(stderr,"PUNCH/PRINT/SUBMIT: write of bitnet spool file failed!\n");
	      exit(EX_TEMPFAIL);
	    }
	    ++reccount;
	  }
	}
	FileSize = ftell(fd); /* Get its size in bytes */

	fflush(fd);
	fseek(fd, recpos, 0); /* Into the old place */
	fprintf(fd, "REC: %8ld",reccount);

	fclose(fd);

	sprintf(FileName, "%s/BIT%ld", BITNET_QUEUE, (long)stats.st_ino);
	rename(tFileName,FileName);

	/* Queue it to the NJE emulator */
	rc = submit_file(FileName, FileSize);
	if (rc)
	  unlink(FileName);
	exit (rc);
}
